### PR TITLE
Issue 6526: LTS - Implement ChunkedSegmentStorage::listSegments

### LINKED ISSUE #6526: LTS - Implement ChunkedSegmentStorage::listSegments
**Is your feature request related to a problem? Please describe.**
ChunkedSegmentStorage::listSegments is not implemented

**Describe the solution you'd like**
Implement ChunkedSegmentStorage::listSegments

**Describe alternatives you've considered**
This is required for implementing admin functionality

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by sachin-j-joshi**
Signed-off-by: Sachin Joshi <sachin.joshi@emc.com>

**Change log description**  
Issue 6526: LTS - Implement ChunkedSegmentStorage::listSegments

**Purpose of the change**  
Fixes #6526 

**What the code does**  
- Make `Storage::listSegments()` return `CompletableFuture`
- Add `getAllEntries() `and `getAllKeys()` methods to `ChunkMetadataStoare`
- Implement `ChunkedSegmentStorage::listSegments`
- Update in memory table mock used in chunked segment storage tests to implement `getAllEntries() `and `getAllKeys()`

**How to verify it**  
All tests should pass.

### CODE REVIEW

File: segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkedSegmentStorage.java [hunk 1]
```diff
@@ -600,9 +600,29 @@ public boolean supportsAtomicWrites() {
...
+     * This method looks only at the segment metadata committed to the underlying {@link ChunkMetadataStore}.
+     *
+     * @return A CompletableFuture that, when completed, will contain an {@link Iterator} that can be used to enumerate and retrieve properties of all the segments.
+     * If the operation failed, it will contain the cause of the failure.
+     */
     @Override
-    public Iterator<SegmentProperties> listSegments() {
-        throw new UnsupportedOperationException("listSegments is not yet supported");
+    public CompletableFuture<Iterator<SegmentProperties>> listSegments() {
+        return metadataStore.getAllEntries()
```
Comment (RaulGracia) [inline segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkedSegmentStorage.java:612]: How many entries could be "all"? How long can be the list of `SegmentProperties` be?

File: segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/BaseMetadataStore.java [hunk 2]
```diff
@@ -775,6 +776,23 @@ long getBufferCount() {
      */
     abstract protected CompletableFuture<Void> writeAll(Collection<TransactionData> dataList);
 
+
+    /**
+     * Retrieve all entries.
```
Comment (RaulGracia) [inline segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/BaseMetadataStore.java:781]: You mean all entries from the `storage_metadata` table segment? If so, please provide a bit more of detail in the description of both methods.

File: segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/TableBasedMetadataStore.java [hunk 3]
```diff
@@ -196,6 +203,56 @@ public TableBasedMetadataStore(String tableName, TableStore tableStore, ChunkedS
...
+    /**
+     * Retrieve all entries.
+     *
+     * @return A CompletableFuture that, when completed, will contain {@link Stream} of {@link StorageMetadata} entries.
+     * If the operation failed, it will be completed with the appropriate exception.
+     */
+    @Override
+    public CompletableFuture<Stream<StorageMetadata>> getAllEntries() {
+        return this.tableStore.entryIterator(tableName, IteratorArgs.builder().fetchTimeout(timeout).build())
+                .thenApplyAsync( i -> getStreamFromTableIterator(i).map(td -> td.getValue()), getExecutor())
```
Comment (RaulGracia) [inline segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/TableBasedMetadataStore.java:215]: Nit: extra sapce.

File: segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/TableBasedMetadataStore.java [hunk 4]
```diff
@@ -196,6 +203,56 @@ public TableBasedMetadataStore(String tableName, TableStore tableStore, ChunkedS
...
+    /**
+     * Retrieve all keys.
+     *
+     * @return A CompletableFuture that, when completed, will contain {@link Stream} of  {@link String} keys.
+     * If the operation failed, it will be completed with the appropriate exception.
+     */
+    @Override
+    public CompletableFuture<Stream<String>> getAllKeys() {
+        return this.tableStore.entryIterator(tableName, IteratorArgs.builder().fetchTimeout(timeout).build())
+                .thenApplyAsync( i -> getStreamFromTableIterator(i).map(td -> td.getKey()), getExecutor())
```
Comment (RaulGracia) [inline segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/TableBasedMetadataStore.java:231]: Nit: extra space.

File: segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/TableBasedMetadataStore.java [hunk 5]
```diff
@@ -196,6 +203,56 @@ public TableBasedMetadataStore(String tableName, TableStore tableStore, ChunkedS
...
+                });
+    }
+
+    /**
+     * Converts the given {@link AsyncIterator} to {@link Stream} of {@link TransactionData}.
+     * @param iterator Iterator to convert.
+     * @return {@link Stream} of {@link TransactionData}.
+     */
+    Stream<TransactionData> getStreamFromTableIterator(AsyncIterator<IteratorItem<TableEntry>> iterator) {
+        return StreamSupport.stream(Spliterators.spliteratorUnknownSize(iterator.asIterator(), Spliterator.ORDERED), true)
```
Comment (RaulGracia) [inline segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/TableBasedMetadataStore.java:244]: What is the expected ordering of this Stream (if any)? Using the "parallel" flag may imply that ordering is going to be hard to guarantee upon consumption. Is ordering actually needed for the Table Entries read as a result?

Comment (RaulGracia) [reply segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/TableBasedMetadataStore.java:244]: If order is not guaranteed, just note that in the description of the method please.

**Review by RaulGracia — COMMENTED** [review]
Looks good, just a couple of comments.

File: segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/BaseMetadataStore.java (hunk 2 above)
Comment (sachin-j-joshi) [reply segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/BaseMetadataStore.java:781]: TableBasedMetadataStore is just one implementation of `ChunkMetadataStore` interface, there could be other implementations of metadata store backed by say ZK, data base or anything that can store KV pairs and has version based optimistic concurrency . 
The `storage_metadata` is specific to implementation  that uses internal tables.

But i get your point. I'll update the description to add more info.

File: segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/TableBasedMetadataStore.java (hunk 5 above)
Comment (sachin-j-joshi) [reply segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/TableBasedMetadataStore.java:244]: Technically we don't really care about ordering. 
I'll change that.

File: segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkedSegmentStorage.java [hunk 6]
```diff
@@ -600,9 +600,29 @@ public boolean supportsAtomicWrites() {
...
+                .thenApplyAsync( storageMetadataStream ->
```
[the remark this answers is not in the document]
Comment (sachin-j-joshi) [reply segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkedSegmentStorage.java]: fixed.

File: segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/BaseMetadataStore.java (hunk 2 above)
Comment (sachin-j-joshi) [reply segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/BaseMetadataStore.java]: fixed.

File: segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/TableBasedMetadataStore.java (hunk 4 above)
Comment (sachin-j-joshi) [reply segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/TableBasedMetadataStore.java]: fixed.

File: segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/TableBasedMetadataStore.java (hunk 5 above)
Comment (sachin-j-joshi) [reply segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/TableBasedMetadataStore.java]: fixed.

File: segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/TableBasedMetadataStore.java (hunk 3 above)
Comment (sachin-j-joshi) [reply segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/TableBasedMetadataStore.java]: fixed.

**Review by RaulGracia — APPROVED** [review]
LGTM.

**Review by anirudhkovuru — APPROVED** [review]
LGTM

**Pull request merged**